### PR TITLE
[#823] Model build operator

### LINKED ISSUE #823: Use K8S resources for model trainings
We need to:

- [ ] Remove small Jenkins
- [ ] Create CRD for repository information (Repo URL, default branch, git key for SSH connection)
- [ ] Create CRD for training job (Repository info, hyper-parameters, resources)
- [ ] Use modern capture mechanism from #820 
- [ ] Create mechanism for capturing metrics and adding them into TrainingJob status with sending to prometheus

### PR DESCRIPTION
**Pull request by mcdoker18**
* Remove jenkins from helm charts
* Remove the core helm chart because it is empty
* Added operator which only build a model
* Added docker containers for the build and operator containers
* Now model is built during e2e tests instead of Jenkins build phase
* Replaced Jenkisfiles in examples with model training crd

This closes #823

### CODE REVIEW

File: containers/operator/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,79 @@
+#########################################################
```
Comment (alexander-semenets) [inline containers/operator/Dockerfile]: Copyright is missed.

File: containers/operator/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+COPY legion/operator/Gopkg.toml legion/operator/Gopkg.lock ./
+RUN dep ensure -vendor-only
+
+COPY legion/operator/pkg/    pkg/
+COPY legion/operator/cmd/    cmd/
+
+RUN GOOS=linux GOARCH=amd64 go build -a -o operator github.com/legion-platform/legion/legion/operator/cmd/manager
+RUN GOOS=linux GOARCH=amd64 go build -a -o builder github.com/legion-platform/legion/legion/operator/cmd/builder
+
+#########################################################
```
Comment (alexander-semenets) [inline containers/operator/Dockerfile:35]: Can we put all Dockerfiles to a common view please.

File: containers/operator/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+COPY legion/cli/Pipfile legion/cli/Pipfile.lock "${LEGION_DIR}/cli/"
+RUN  cd "${LEGION_DIR}/cli" && pipenv install --system --three
+
+COPY legion/sdk "${LEGION_DIR}/sdk"
+COPY legion/cli "${LEGION_DIR}/cli"
+
+RUN pip install --no-deps -e "${LEGION_DIR}/sdk" && \
+    pip install --no-deps -e "${LEGION_DIR}/cli"
+
+# TODO: fixed
```
Comment (alexander-semenets) [inline containers/operator/Dockerfile]: abandoned todo

File: containers/pipeline-agent/Dockerfile [hunk 4]
```diff
@@ -31,6 +31,9 @@ RUN curl -fsSLO https://kubernetes-helm.storage.googleapis.com/helm-${HELM_VERSI
     tar xzf /tmp/helm/helm-${HELM_VERSION}-linux-amd64.tar.gz -C /tmp/helm && \
     mv /tmp/helm/linux-amd64/helm /usr/local/bin/helm && rm -rf /tmp/helm
 
+# Used for robot tests
```
Comment (alexander-semenets) [inline containers/pipeline-agent/Dockerfile:50]: copyright is missed

File: containers/operator/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+RUN ssh-keyscan github.com > /etc/ssh/ssh_known_hosts
+
+COPY --from=builder /go/src/github.com/legion-platform/legion/legion/operator/builder "${LEGION_DIR}/"
+WORKDIR "${LEGION_DIR}"
+
+CMD ["./builder"]
+
+#########################################################
+#########################################################
+#########################################################
```
Comment (alexander-semenets) [inline containers/operator/Dockerfile:81]: Some comments about each image purpose would be useful

File: containers/operator/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+RUN  cd "${LEGION_DIR}/cli" && pipenv install --system --three
+
+COPY legion/sdk "${LEGION_DIR}/sdk"
+COPY legion/cli "${LEGION_DIR}/cli"
+
+RUN pip install --no-deps -e "${LEGION_DIR}/sdk" && \
+    pip install --no-deps -e "${LEGION_DIR}/cli"
+
+# TODO: fixed
+RUN ssh-keyscan github.com > /etc/ssh/ssh_known_hosts
```
Comment (alexander-semenets) [inline containers/operator/Dockerfile]: Do we really want to freeze the key in image? Shouldn't it be better set up at the place where we do checkout dynamically? I guess the keys may be update.

File: legion/operator/Makefile [hunk 7]
```diff
@@ -0,0 +1,57 @@
+
+# Image URL to use all building/pushing image targets
+IMG ?= legion/operator:latest
```
Comment (alexander-semenets) [inline legion/operator/Makefile]: for prod and stage releases we use legionplatform and legionplatformtest registries respectively.
Wouldn't it be a problem if we have legion/ registry strictly defined here?

File: legion/operator/Makefile [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+LEGION_SRC_ROOT = ../../
+CONTAINERS = ${LEGION_SRC_ROOT}/build/containers
+TARGET_HELM_CHART = ${LEGION_SRC_ROOT}/deploy/helms/legion/
+TARGET_YAML_STORAGE = ${TARGET_HELM_CHART}/templates/operator/generated
```
Comment (alexander-semenets) [inline legion/operator/Makefile:6]: I guess those paths /build and /deploy are not valid any more?

File: legion/operator/pkg/builder/builder.go [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+	pod, err := mb.clientset.CoreV1().Pods(mb.config.Namespace).Get(mb.config.PodName, metav1.GetOptions{})
+
+	if err != nil {
+		log.Error(err, "Can't get pod %s", pod.Name)
+		return "", err
+	}
+
+	for _, container := range pod.Status.ContainerStatuses {
+		if container.Name == "model" {
+			return container.ContainerID[len("docker://"):], nil
```
Comment (kirillmakhonin) [inline legion/operator/pkg/builder/builder.go]: Please use constants

File: legion/operator/pkg/utils/git.go [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+	if err != nil {
+		return
+	}
+
+	_ = os.RemoveAll(cloneDir)
+
+	repository, err := git.PlainClone(cloneDir, false, &git.CloneOptions{
+		URL:      repositoryUrl,
+		Progress: os.Stdout,
+		Auth:     gitAuth,
```
Comment (kirillmakhonin) [inline legion/operator/pkg/utils/git.go:44]: Does it work with https://... Git URLs?

File: legion/operator/pkg/utils/git.go [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+
+	return nil
+}
+
+// We assume that ssh key located in user home directory as %HOME_DIR/.ssh/id_rsa
+func getSshKeyAuth() (transport.AuthMethod, error) {
+	var auth transport.AuthMethod
+
+	// TODO: make location configurable
+	privateSshKeyFile := "/home/root/.ssh/id_rsa"
```
Comment (kirillmakhonin) [inline legion/operator/pkg/utils/git.go]: Looks like this should be ENV configurable.

**Review by kirillmakhonin — CHANGES_REQUESTED** [review]

File: legion/operator/Makefile (hunk 7 above)
Comment (mcdoker18) [reply legion/operator/Makefile]: Removed the variable and target

File: containers/operator/Dockerfile (hunk 2 above)
Comment (mcdoker18) [reply containers/operator/Dockerfile:35]: Please, explain what do you mean.

**Review by kirillmakhonin — DISMISSED** [review]

**Comment by mcdoker18** [comment]
bot, test app

**Comment by mcdoker18** [comment]
bot, test all

**Comment by mcdoker18** [comment]
bot, test app

**Comment by mcdoker18** [comment]
bot, test app

**Comment by mcdoker18** [comment]
bot, test app

**Comment by mcdoker18** [comment]
bot, test app

**Comment by mcdoker18** [comment]
bot, test app

**Comment by mcdoker18** [comment]
bot, test app

**Review by kirillmakhonin — APPROVED** [review]

**Review by alexander-semenets — APPROVED** [review]

**Pull request merged**